### PR TITLE
Feature/gem 1353

### PR DESCRIPTION
**Pull request by gesterzhou**

### CODE REVIEW

**Comment by gesterzhou** [comment]
This is the first draft of pull request for review. It's opt to change.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java [hunk 1]
```diff
@@ -1261,7 +1333,14 @@ void basicUpdateEntryVersion(EntryEventImpl event) throws EntryNotFoundException
   }
 
   protected void distributeUpdateEntryVersionOperation(EntryEventImpl event) {
-    new UpdateEntryVersionOperation(event).distribute();
+    UpdateEntryVersionOperation op = new UpdateEntryVersionOperation(event);
+    long viewVersion = -1;
+    try {
+      viewVersion = op.startOperation();
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java]: For cases like this one it would be nice to move the try/finally to a new method in DistributedCacheOperation and just invoke that method.  Maybe call it distribute()???

```
public void distributed() {
  long token;
  try {
    token = startOperation();
    _distribute(); // private method
  } finally {
    endOperation(token);
  }
}
```
If you do this you can revert the changes you've made to a lot of classes.

Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java]: Out of curiosity, what happens if the startOperation() method returns a -1 (I think that may be possible based on how it's currently written or if an exception is thrown before token is assigned.  What is the expected way to handle endOperation with an invalid/-1 token?

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedCacheOperation.java [hunk 2]
```diff
@@ -240,12 +240,37 @@ public boolean containsRegionContentChange() {
...
+    if (logger.isTraceEnabled(LogMarker.STATE_FLUSH_OP)) {
+      logger.trace(LogMarker.STATE_FLUSH_OP, "dispatching operation in view version {}",
+          viewVersion);
+    }
+    return viewVersion;
+  }
+
+  public void endOperation(long viewVersion) {
+    DistributedRegion region = getRegion();
+    if (viewVersion != -1) {
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/DistributedCacheOperation.java]: If somehow we get a -1, will the state flush ever "end?" or are we stuck?
